Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,684 @@
+  auto ylf = m_yvalues[(y * w) + x + w];
+  auto yrf = m_yvalues[(y * w) + x + w + 1];
+  auto yrb = m_yvalues[(y * w) + x + 1];
+
+  // calculate midpoint of trapeziod
+  double a = std::abs((double)(xrf - xlf));
+  double b = std::abs((double)(xrb - xlb));
+  double h = std::abs((double)(ylb - ylf));
+  auto cx = ((a + b) / 4);
+  auto cy = h / 2;
+
+  // store detector position before translating to origin
+  auto xpos = xlb + cx;
+  auto ypos = ylb + cy;
+
+  // Translate detector shape to origin
+  xlf -= xpos;
+  xrf -= xpos;
+  xrb -= xpos;
+  xlb -= xpos;
+  ylf -= ypos;
+  yrf -= ypos;
+  yrb -= ypos;
+  ylb -= ypos;
+
+  boost::shared_ptr<Mantid::Geometry::Object> shape =
+      streamShape(name, xlb, xlf, xrf, xrb, ylb, ylf, yrf, yrb);
+
+  // Create detector
+  auto detector = new Detector(name, id, shape, parent);
+
+  // Set detector position relative to parent
+  V3D pos(xpos, ypos, 0);
+
+  detector->translate(pos);
+
+  return detector;
+}
+
+/** Returns the minimum detector id
+* @return minimum detector id
+*/
+int StructuredDetector::minDetectorID() {
+  if (m_map)
+    return m_base->m_minDetId;
+  return m_minDetId;
+}
+
+//-------------------------------------------------------------------------------------------------
+/** Returns the maximum detector id
+* @return maximum detector id
+*/
+int StructuredDetector::maxDetectorID() {
+  if (m_map)
+    return m_base->m_maxDetId;
+  return m_maxDetId;
+}
+
+boost::shared_ptr<const IComponent>
+StructuredDetector::getComponentByName(const std::string &cname,
+                                       int nlevels) const {
+  // exact matches
+  if (cname == this->getName())
+    return boost::shared_ptr<const IComponent>(this);
+
+  // cache the detector's name as all the other names are longer
+  // The extra ( is because all children of this have that as the next character
+  // and this prevents Bank11 matching Bank 1
+  const std::string MEMBER_NAME = this->getName() + "(";
+
+  // check that the searched for name starts with the detector's
+  // name as they are generated
+  if (cname.substr(0, MEMBER_NAME.length()).compare(MEMBER_NAME) != 0) {
+    return boost::shared_ptr<const IComponent>();
+  } else {
+    return CompAssembly::getComponentByName(cname, nlevels);
+  }
+}
+
+//-------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------
+// ------------ IObjComponent methods ----------------
+//-------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------------------------
+/// Does the point given lie within this object component?
+bool StructuredDetector::isValid(const V3D &) const {
```

Comment:
No other reason than they aren't implemented in the RectangularDetector. I thought these were not implemented because the detector surface is assumed to be infinitely thin (Same for the current implementation of the StructuredDetector)